File: c/src/proto/data_audit_chain.c
#include "auth/data_audit_chain.h"

#include <limits.h>
#include <sodium.h>
#include <string.h>

static const uint8_t DOMAIN[] = "ZKARCHE-DATA-AUDIT-v1";

static void update_u64(crypto_hash_sha256_state *state, uint64_t value) {
    uint8_t be[8];
    size_t i;
    for (i = 0u; i < sizeof be; ++i) {
        be[sizeof be - 1u - i] = (uint8_t)(value & 0xffu);
        value >>= 8;
    }
    (void)crypto_hash_sha256_update(state, be, sizeof be);
}

void data_audit_state_genesis(data_audit_state_t *state) {
    if (state == NULL) {
        return;
    }
    memset(state->head, 0, sizeof state->head);
    state->next_sequence = 1u;
}

data_audit_result_t data_audit_entry_digest(
    uint8_t out[32],
    const uint8_t previous_head[32],
    uint64_t sequence,
    const data_audit_event_t *event)
{
    crypto_hash_sha256_state state;
    uint8_t kind;

    if (out == NULL || previous_head == NULL || event == NULL) {
        return DATA_AUDIT_INVALID_ARGUMENT;
    }
    if (sequence == 0u) {
        return DATA_AUDIT_INVALID_SEQUENCE;
    }
    if (event->kind != DATA_AUDIT_EVENT_RELEASE_GRANTED
        && event->kind != DATA_AUDIT_EVENT_RELEASE_DENIED) {
        return DATA_AUDIT_INVALID_ARGUMENT;
    }

    kind = (uint8_t)event->kind;
    (void)crypto_hash_sha256_init(&state);
    (void)crypto_hash_sha256_update(&state, DOMAIN, sizeof DOMAIN - 1u);
    (void)crypto_hash_sha256_update(&state, previous_head, 32u);
    update_u64(&state, sequence);
    (void)crypto_hash_sha256_update(&state, &kind, 1u);
    (void)crypto_hash_sha256_update(&state, event->release_context_hash, 32u);
    update_u64(&state, event->authorization_generation);
    update_u64(&state, event->revocation_epoch);
    update_u64(&state, event->policy_epoch);
    (void)crypto_hash_sha256_final(&state, out);
    return DATA_AUDIT_OK;
}

data_audit_result_t data_audit_append(
    data_audit_state_t *state,
    const data_audit_event_t *event,
    uint8_t out[32])
{
    data_audit_result_t rc;
    uint8_t digest[32];

    if (state == NULL || event == NULL || out == NULL) {
        return DATA_AUDIT_INVALID_ARGUMENT;
    }
    if (state->next_sequence == UINT64_MAX) {
        return DATA_AUDIT_SEQUENCE_EXHAUSTED;
    }

    rc = data_audit_entry_digest(digest, state->head, state->next_sequence, event);
    if (rc != DATA_AUDIT_OK) {
        return rc;
    }
    memcpy(state->head, digest, sizeof state->head);
    memcpy(out, digest, sizeof digest);
    state->next_sequence += 1u;
    return DATA_AUDIT_OK;
}

bool data_audit_verify_transition(
    const uint8_t previous_head[32],
    uint64_t sequence,
    const data_audit_event_t *event,
    const uint8_t expected_head[32])
{
    uint8_t actual[32];
    if (expected_head == NULL) {
        return false;
    }
    if (data_audit_entry_digest(actual, previous_head, sequence, event) != DATA_AUDIT_OK) {
        return false;
    }
    return sodium_memcmp(actual, expected_head, sizeof actual) == 0;
}

File: rust/crates/proto/src/data_audit_chain.rs
//! Bounded local audit hash chain for protected DATA release events.
//!
//! This module is wire-neutral. Audit state is evidence about local release
//! decisions; it is never an authorization source and never substitutes for
//! current AUTH, revocation, lineage, policy, or device release authority.

use sha2::{Digest, Sha256};

const DOMAIN: &[u8] = b"ZKARCHE-DATA-AUDIT-v1";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum DataAuditEventKind {
    ReleaseGranted = 1,
    ReleaseDenied = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DataAuditEvent {
    /// Commitment to the canonical release context; never protected plaintext.
    pub release_context_hash: [u8; 32],
    pub authorization_generation: u64,
    pub revocation_epoch: u64,
    pub policy_epoch: u64,
    pub kind: DataAuditEventKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct DataAuditState {
    pub head: [u8; 32],
    pub next_sequence: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DataAuditError {
    InvalidSequence,
    SequenceExhausted,
}

impl DataAuditState {
    pub const fn genesis() -> Self {
        Self {
            head: [0u8; 32],
            next_sequence: 1,
        }
    }
}

pub fn data_audit_entry_digest(
    previous_head: &[u8; 32],
    sequence: u64,
    event: &DataAuditEvent,
) -> Result<[u8; 32], DataAuditError> {
    if sequence == 0 {
        return Err(DataAuditError::InvalidSequence);
    }

    let mut h = Sha256::new();
    h.update(DOMAIN);
    h.update(previous_head);
    h.update(sequence.to_be_bytes());
    h.update([event.kind as u8]);
    h.update(event.release_context_hash);
    h.update(event.authorization_generation.to_be_bytes());
    h.update(event.revocation_epoch.to_be_bytes());
    h.update(event.policy_epoch.to_be_bytes());
    Ok(h.finalize().into())
}

pub fn append_data_audit_event(
    state: &mut DataAuditState,
    event: &DataAuditEvent,
) -> Result<[u8; 32], DataAuditError> {
    if state.next_sequence == u64::MAX {
        return Err(DataAuditError::SequenceExhausted);
    }
    let digest = data_audit_entry_digest(&state.head, state.next_sequence, event)?;
    state.head = digest;
    state.next_sequence += 1;
    Ok(digest)
}

pub fn verify_data_audit_transition(
    previous_head: &[u8; 32],
    sequence: u64,
    event: &DataAuditEvent,
    expected_head: &[u8; 32],
) -> bool {
    data_audit_entry_digest(previous_head, sequence, event)
        .map(|actual| actual == *expected_head)
        .unwrap_or(false)
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXPECTED_FIRST: [u8; 32] = [
        0x9d, 0xf3, 0x6e, 0x00, 0xb5, 0x56, 0x84, 0xfc,
        0xd1, 0xaf, 0x82, 0x97, 0xcb, 0x83, 0x2f, 0x61,
        0x6b, 0x62, 0xcd, 0xe5, 0x18, 0x1c, 0x0c, 0xe9,
        0xf2, 0xcd, 0x8e, 0x99, 0x7b, 0xe5, 0x80, 0x98,
    ];

    fn event() -> DataAuditEvent {
        let mut release_context_hash = [0u8; 32];
        for (i, byte) in release_context_hash.iter_mut().enumerate() {
            *byte = i as u8;
        }
        DataAuditEvent {
            release_context_hash,
            authorization_generation: 7,
            revocation_epoch: 11,
            policy_epoch: 13,
            kind: DataAuditEventKind::ReleaseGranted,
        }
    }

    #[test]
    fn canonical_first_entry_matches_cross_language_fixture() {
        let mut state = DataAuditState::genesis();
        assert_eq!(append_data_audit_event(&mut state, &event()).unwrap(), EXPECTED_FIRST);
        assert_eq!(state.head, EXPECTED_FIRST);
        assert_eq!(state.next_sequence, 2);
    }

    #[test]
    fn continuity_and_event_mutation_are_detectable() {
        let e = event();
        assert!(verify_data_audit_transition(&[0u8; 32], 1, &e, &EXPECTED_FIRST));
        assert!(!verify_data_audit_transition(&[1u8; 32], 1, &e, &EXPECTED_FIRST));
        assert!(!verify_data_audit_transition(&[0u8; 32], 2, &e, &EXPECTED_FIRST));

        let mut changed = e;
        changed.release_context_hash[0] ^= 1;
        assert!(!verify_data_audit_transition(&[0u8; 32], 1, &changed, &EXPECTED_FIRST));

        changed = e;
        changed.authorization_generation += 1;
        assert!(!verify_data_audit_transition(&[0u8; 32], 1, &changed, &EXPECTED_FIRST));

        changed = e;
        changed.revocation_epoch += 1;
        assert!(!verify_data_audit_transition(&[0u8; 32], 1, &changed, &EXPECTED_FIRST));

        changed = e;
        changed.policy_epoch += 1;
        assert!(!verify_data_audit_transition(&[0u8; 32], 1, &changed, &EXPECTED_FIRST));

        changed = e;
        changed.kind = DataAuditEventKind::ReleaseDenied;
        assert!(!verify_data_audit_transition(&[0u8; 32], 1, &changed, &EXPECTED_FIRST));
    }

    #[test]
    fn sequence_zero_and_exhaustion_fail_closed() {
        assert_eq!(
            data_audit_entry_digest(&[0u8; 32], 0, &event()),
            Err(DataAuditError::InvalidSequence)
        );
        let mut state = DataAuditState {
            head: [0u8; 32],
            next_sequence: u64::MAX,
        };
        assert_eq!(
            append_data_audit_event(&mut state, &event()),
            Err(DataAuditError::SequenceExhausted)
        );
    }
}
